### PR TITLE
Introduce intrinsic stable hash function Fixes #4986

### LINKED ISSUE #4986: Add property function GetStableHash for String
Some projects wish to include hashes of strings (often paths) in output folder names, which must be defined at evaluation time. [`System.String.GetHashCode`](https://docs.microsoft.com/dotnet/api/system.string.gethashcode?view=netframework-4.8) is available but unsuitable for many desirable purposes: it's documented to vary across 64- and 32-bit .NET implementations and can also vary between .NET Framework and .NET Core.

MSBuild has an internal method to get a stable hash that's currently used for IPC

https://github.com/microsoft/msbuild/blob/fa773bb8d44a358963481c02e772509dc408a6d9/src/Shared/CommunicationsUtilities.cs#L673-L707

We should expose it as `$([MSBuild]::StableStringHash())`.

Open questions:
* Name
* Should we override calls to `String.GetHashCode()`?

### PR DESCRIPTION
**Pull request by Forgind**
Fixes #4986 by introducing a new, stable hash function. Takes the safe route of not overwriting the previous hash function.

### CODE REVIEW

**Comment by cdmihai** [comment]
Update the docs as well: https://docs.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2019#msbuild-property-functions

**Review by cdmihai — APPROVED** [review]

**Review by BenVillalobos — APPROVED** [review]
LGTM!

File: src/Build.UnitTests/Evaluation/Expander_Tests.cs [hunk 1]
```diff
@@ -3609,6 +3609,38 @@ public void PropertyStringConstructorConsumingItemMetadata(string metadatumName,
...
+                (int)expander.ExpandPropertiesLeaveTypedAndEscaped("$([MSBuild]::StableStringHash('cut1s'))", ExpanderOptions.ExpandProperties, MockElementLocation.Instance),
+                (int)expander.ExpandPropertiesLeaveTypedAndEscaped("$([MSBuild]::StableStringHash('cat1so'))", ExpanderOptions.ExpandProperties, MockElementLocation.Instance),
+                (int)expander.ExpandPropertiesLeaveTypedAndEscaped("$([MSBuild]::StableStringHash('cats1'))", ExpanderOptions.ExpandProperties, MockElementLocation.Instance),
+                (int)expander.ExpandPropertiesLeaveTypedAndEscaped("$([MSBuild]::StableStringHash('acat1s'))", ExpanderOptions.ExpandProperties, MockElementLocation.Instance),
+                (int)expander.ExpandPropertiesLeaveTypedAndEscaped("$([MSBuild]::StableStringHash('cat12s'))", ExpanderOptions.ExpandProperties, MockElementLocation.Instance),
+                (int)expander.ExpandPropertiesLeaveTypedAndEscaped("$([MSBuild]::StableStringHash('cat1s'))", ExpanderOptions.ExpandProperties, MockElementLocation.Instance)
```
Comment (ladipro) [inline src/Build.UnitTests/Evaluation/Expander_Tests.cs:3626]: nit: Perhaps you could create an array with just the strings to hash (`"cat1s", "cat1z", ...`) to eliminate code/text duplication.

**Review by ladipro — APPROVED** [review]

**Review by rainersigwald — APPROVED** [review]

File: src/Shared/CommunicationsUtilities.cs [hunk 2]
```diff
@@ -83,6 +83,7 @@ internal Handshake(HandshakeOptions nodeType)
             CommunicationsUtilities.Trace("Handshake salt is " + handshakeSalt);
             string toolsDirectory = (nodeType & HandshakeOptions.X64) == HandshakeOptions.X64 ? BuildEnvironmentHelper.Instance.MSBuildToolsDirectory64 : BuildEnvironmentHelper.Instance.MSBuildToolsDirectory32;
             CommunicationsUtilities.Trace("Tools directory is " + toolsDirectory);
+            salt = CommunicationsUtilities.GetHashCode(handshakeSalt + toolsDirectory);
             salt = CommunicationsUtilities.GetHandshakeHashCode(handshakeSalt + toolsDirectory);
```
Comment (Forgind) [inline src/Shared/CommunicationsUtilities.cs:87]:
```suggestion
```

**Pull request merged**